### PR TITLE
[REEF-1778] Ensure ITask.Dispose() is executed

### PR DESCRIPTION
**Pull request by markusweimer**
This moves the reporting of the result of a task *after* the call to `Dispose` on that task to ensure execution of that method.

To do so, I made the calls to the task start and stop handlers explicit, and kept track of the exceptions thrown and need to send a result in the `TaskRuntime`.

Also, this fixes a bug in `TaskRuntimeTests` which did not wait for the spawned thread to finish before validation of the test results.

JIRA:
  [REEF-1778](https://issues.apache.org/jira/browse/REEF-1778)

### CODE REVIEW

**Comment by markusweimer** [comment]
@jwang98052, can you have a look at this?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs [hunk 1]
```diff
@@ -93,54 +93,81 @@ public Thread StartTaskOnNewThread()
...
                 }
+
+                // Inform the driver about the result.
+                if (returnResultToDriver)
+                {   
+                    _currentStatus.SetResult(resultReturnedByTask);
+                }
+
+                // If the ITask.Dispose() method threw an Exception, crash the Evaluator.
+                if (exceptionThrownByTaskDispose != null)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: So if Dispose throws exception and task does get result, it still sets result first, then  throw exception to crash the evaluator?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs [hunk 2]
```diff
@@ -93,54 +93,81 @@ public Thread StartTaskOnNewThread()
...
+                bool returnResultToDriver = true;
+
+                // Exception thrown during `Dispose`, if any.
+                Exception exceptionThrownByTaskDispose = null;
+
                 try
                 {
+                    // Run the handlers for `TaskStart`
                     Logger.Log(Level.Verbose, "Set running status for task");
+                    _currentStatus.RunTaskStartHandlers();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: The original _taskLifeCycle.Start(0 was in the _heartBeatManager lock. Do you on purposely take it out? Any reason?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs [hunk 3]
```diff
@@ -93,54 +93,81 @@ public Thread StartTaskOnNewThread()
...
+                    // Call the Task
                     Logger.Log(Level.Verbose, "Calling into user's task.");
-                    var result = _userTask.Call(null);
+                    resultReturnedByTask = _userTask.Call(null);
                     Logger.Log(Level.Info, "Task Call Finished");
-                    _currentStatus.SetResult(result);
 
-                    const Level resultLogLevel = Level.Verbose;
+                    // Run the handlers for `TaskStop`
+                    _currentStatus.RunTaskStopHandlers();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: Same for  _taskLifeCycle.Stop()

**Comment by jwang98052** [comment]
I completed a pass.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: Yes, that is the behavior of the code prior to this change. I don't want to change semantics in this change. Once our thread on the mailing list converges, we can make changes like that.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: No, this was an omission. Thanks for pointing this out! Would it be enough to add the lock to the `RunTaskStartHandlers()` method or do we need to acquire the lock here at the callsite?

**Comment by markusweimer** [comment]
@jwang98052 Thanks! I answered :)

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs]: On second thought, the processing of the event handlers doesn't need to lock the heartbeat manager. There is no event inside that could trigger a heartbeat.

**Comment by markusweimer** [comment]
I looked into the AppVeyor failure. It seems unrelated to the changes made here.

**Comment by markusweimer** [comment]
Rebased the branch.

**Comment by jwang98052** [comment]
I will test and merge

**Pull request closed without merging**